crates/l1tx/src/envelope/parser.rs

```diff
@@ -19,75 +24,85 @@ pub enum EnvelopeParseError {
 ///
 /// # Errors
 ///
 /// This function errors if it cannot parse the [`L1Payload`]
-pub fn parse_envelope_data(
+pub fn parse_envelope_payloads(
     script: &ScriptBuf,
-    rollup_name: &str,
-) -> Result<L1Payload, EnvelopeParseError> {
+    params: &RollupParams,
+) -> Result<Vec<L1Payload>, EnvelopeParseError> {
     let mut instructions = script.instructions();
 
-    enter_envelope(&mut instructions)?;
-    // Parse name
-    let (tag, name) = parse_bytes_pair(&mut instructions)?;
+    let mut payloads = Vec::new();
+    // TODO: make this sophisticated, i.e. even if one payload parsing fails, continue finding other
+    // envelopes and extracting payloads. Or is that really necessary?
+    while let Ok(payload) = parse_l1_payload(&mut instructions, params) {
+        payloads.push(payload);
+    }
+    Ok(payloads)
+}
 
-    let extracted_rollup_name = match (tag, name) {
-        (ROLLUP_NAME_TAG, namebytes) => {
-            String::from_utf8(namebytes.to_vec()).map_err(|_| EnvelopeParseError::InvalidNameValue)
-        }
-        _ => Err(EnvelopeParseError::InvalidNameTag),
-    }?;
+fn parse_l1_payload(
+    instructions: &mut Instructions,
+    params: &RollupParams,
+) -> Result<L1Payload, EnvelopeParseError> {
+    enter_envelope(instructions)?;
 
-    if extracted_rollup_name != rollup_name {
-        return Err(EnvelopeParseError::InvalidNameTag);
-    }
+    // Parse type
+    let ptype = next_bytes(instructions)
+        .and_then(|bytes| parse_payload_type(bytes, params))
+        .ok_or(EnvelopeParseError::InvalidTag)?;
 
     // Parse version
-    let (tag, ver) = parse_bytes_pair(&mut instructions)?;
-    let _version = match (tag, ver) {
-        (VERSION_TAG, [v]) => Ok(v),
-        (VERSION_TAG, _) => Err(EnvelopeParseError::InvalidVersion),
-        _ => Err(EnvelopeParseError::InvalidVersionTag),
-    }?;
-
-    // Parse bytes
-    let tag = next_bytes(&mut instructions).ok_or(EnvelopeParseError::InvalidBlobTag)?;
-    let size = next_int(&mut instructions);
-    match (tag, size) {
-        (BATCH_DATA_TAG, Some(size)) => {
-            let batch_data = extract_n_bytes(size, &mut instructions)?;
-            Ok(L1Payload::new_checkpoint(batch_data)) // TODO: later this will discern checkpoint
-                                                      // and da and any other payload types
-        }
-        (BATCH_DATA_TAG, None) => Err(EnvelopeParseError::InvalidBlob),
-        _ => Err(EnvelopeParseError::InvalidBlobTag),
+    let _version = next_bytes(instructions)
+        .and_then(validate_version)
+        .ok_or(EnvelopeParseError::InvalidVersion)?;
+
+    // Parse size
+    let size = next_u32(instructions).ok_or(EnvelopeParseError::InvalidSize)?;
+    // Parse payload
+    let payload = extract_n_bytes(size, instructions)?;
+    Ok(L1Payload::new(payload, ptype))
+}
+
+fn parse_payload_type(bytes: &[u8], params: &RollupParams) -> Option<L1PayloadType> {
+    let str = from_utf8(bytes).ok()?;
```

Comment:
Instead of trying to parse the bytes as UTF-8 first, interpret the strings we're comparing against as bytes and compare that like `tag_bytes == params.checkpoint_tag.as_bytes()`.

Also, don't like var names that shadow type names, would have called this `tag_str` to also be more explicit.

Reply:
That's a good point. Changed to checkping bytes. Conversion to string was really unnecessary.